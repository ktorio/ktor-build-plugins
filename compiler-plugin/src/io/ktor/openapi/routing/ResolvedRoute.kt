package io.ktor.openapi.routing

data class ResolvedRoute(
    val path: String,
    val method: String,
    val fields: RouteFieldList = emptyList(),
)