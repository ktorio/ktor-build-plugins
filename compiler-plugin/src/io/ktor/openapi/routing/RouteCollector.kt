package io.ktor.openapi.routing

import org.jetbrains.kotlin.fir.FirSession

object RouteCollector {
    fun collectRoutes(graph: RouteCallGraph): List<ResolvedRoute> {
        // Process each route node to generate ResolvedRoutes
        return graph.routes.flatMap { route ->
            if (route.method == null) return@flatMap emptyList()
            
            // Find all paths from roots to this route
            val allPaths = graph.findAllPathsToRoots(route)
            
            // Process each path and merge fields
            with(graph.session) {
                allPaths.mapNotNull { path ->
                    resolve(
                        parents = path,
                        children = graph.getChildren(route)
                    )
                }
            }
        }
    }

    context(session: FirSession)
    fun resolve(
        parents: Iterable<RouteNode> = emptyList(),
        children: Iterable<RouteNode> = emptyList(),
    ): ResolvedRoute? {
        var stack = RouteStack(session)
        var fields = emptyList<RouteField>()
        for (ancestor in parents) {
            fields = ancestor.fields(stack).merge(fields)
            stack += ancestor
        }

        for (child in children) {
            fields = fields.merge(child.fields(stack))
        }

        val path = fields.path ?: return null
        val method = fields.method ?: return null
        if (RouteField.Ignore in fields) return null

        return ResolvedRoute(
            path = path,
            method = method,
            fields = fields,
        )
    }
}