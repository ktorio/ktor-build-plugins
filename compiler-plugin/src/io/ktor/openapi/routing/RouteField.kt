package io.ktor.openapi.routing

import io.ktor.openapi.model.JsonSchema
import io.ktor.openapi.model.JsonType
import kotlinx.serialization.json.JsonElement
import org.jetbrains.kotlin.utils.addToStdlib.firstIsInstanceOrNull

/**
 * Sealed class representing different KDoc parameters for OpenAPI documentation.
 */
sealed interface RouteField {

    fun merge(other: RouteField): RouteField? =
        if (this::class == other::class) this else null

    sealed interface SchemaHolder : RouteField {
        val schema: SchemaReference?
        val attributes: Map<String, JsonElement>
    }

    sealed interface Content : SchemaHolder {
        val contentType: String?
        val description: String?
    }

    sealed interface Parameter : SchemaHolder {
        val name: String
        val description: String?
        val `in`: String
    }

    /**
     * Built into the OpenAPI JSON structure, so ignored when building the path item.
     */
    sealed interface Transient : RouteField

    data class Method(val method: String) : Transient {
        override fun merge(other: RouteField): RouteField? =
            if (other is Method && method == other.method) this else null
    }

    data class Path(val path: String) : Transient {
        override fun merge(other: RouteField): RouteField? =
            if (other is Path)
                Path("${other.path}/${path.removePrefix("/")}")
            else null
    }

    data class Schema(
        val name: String,
        val schema: JsonSchema
    ) : Transient {
        override fun merge(other: RouteField): RouteField? =
            if (other is Schema && other.name == name) this else null
    }

    /**
     * Associates the endpoint with a tag for grouping.
     *
     * Format: `@tag TagName`
     */
    data class Tag(val name: String) : RouteField {
        override fun merge(other: RouteField): RouteField? =
            if (other is Tag && name == other.name) this else null
    }

    /**
     * Describes a path parameter.
     *
     * Format: `@path name description`
     */
    data class PathParam(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "path"
        override fun merge(other: RouteField): RouteField? =
            if (other is PathParam && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Describes a query parameter.
     *
     * Format: `@query name description`
     */
    data class QueryParam(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "query"
        override fun merge(other: RouteField): RouteField? =
            if (other is QueryParam && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Describes a header parameter.
     *
     * Format: `@header name description`
     */
    data class Header(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "header"
        override fun merge(other: RouteField): RouteField? =
            if (other is Header && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
            ) else null
    }

    /**
     * Describes a cookie parameter.
     *
     * Format: `@cookie name description`
     */
    data class Cookie(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "cookie"
        override fun merge(other: RouteField): RouteField? =
            if (other is Cookie && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Documents the request body type.
     *
     * Format: `@body [Type] description`
     */
    data class Body(
        override val contentType: String? = null,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Content {
        override fun merge(other: RouteField): RouteField? =
            if (other is Body) copy(
                contentType = contentType ?: other.contentType,
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Documents a response code with optional type and description.
     *
     * Format: `@response code [Type] description`
     */
    data class Response(
        val code: String,
        override val contentType: String? = null,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Content {
        override fun merge(other: RouteField): RouteField? =
            if (other is Response && code == other.code)
                copy(
                    contentType = contentType ?: other.contentType,
                    schema = schema ?: other.schema,
                    description = description ?: other.description,
                    attributes = attributes + other.attributes,
                )
            else null
    }

    /**
     * Marks an endpoint as deprecated.
     *
     * Format: `@deprecated reason`
     */
    data class Deprecated(val reason: String) : RouteField

    /**
     * Provides a detailed endpoint description.
     *
     * Format: `@description text`
     */
    data class Description(val text: String) : RouteField


    /**
     * Provides a summary of the endpoint.
     * Format: `@summary text`
     */
    data class Summary(val text: String) : RouteField

    /**
     * Documents security requirements.
     *
     * Format: `@security scheme`
     */
    data class Security(val scheme: String) : RouteField {
        override fun merge(other: RouteField): RouteField? =
            if (other is Security && scheme == other.scheme) this else null
    }
}

typealias RouteFieldList = List<RouteField>

val RouteFieldList.path: String?
    get() = firstIsInstanceOrNull<RouteField.Path>()?.path

val RouteFieldList.method: String?
    get() = firstIsInstanceOrNull<RouteField.Method>()?.method

/**
 * Merges two lists of route fields into a new list.
 */
fun RouteFieldList.merge(other: RouteFieldList) = buildList {
    val otherMutable = other.toMutableList()
    for (field in this@merge) {
        val match = otherMutable.indices.firstNotNullOfOrNull { i ->
            field.merge(otherMutable[i])?.also {
                otherMutable.removeAt(i)
                add(it)
            }
        }
        // if no compatible field was found, add this field to the result
        if (match == null) add(field)
    }
    // add all remaining unmerged fields to the result
    // TODO ignore all summary fields because we currently have some false positives
    addAll(otherMutable.filterNot { it is RouteField.Summary })
}

sealed interface SchemaReference {

    fun asSchema(): JsonSchema

    data class Resolved(
        val schema: JsonSchema
    ) : SchemaReference {
        override fun asSchema(): JsonSchema = schema
    }

    sealed interface Link: SchemaReference {
        val name: String

        data class Simple(override val name: String, val jsonType: JsonType) : Link {
            override fun asSchema(): JsonSchema =
                JsonSchema(jsonType)
        }
        data class Reference(override val name: String) : Link {
            override fun asSchema(): JsonSchema =
                JsonSchema(ref = "#/components/schemas/$name")
        }
        data class Array(val element: Link) : Link by element {
            override fun asSchema(): JsonSchema =
                JsonSchema(
                    type = JsonType.array,
                    items = element.asSchema()
                )
        }
        data class Optional(val delegate: Link) : Link by delegate {
            override fun asSchema(): JsonSchema =
                delegate.asSchema()
                    .copy(required = false)
        }
    }
}

fun SchemaReference.hasReference(): Boolean = getReference() != null

fun SchemaReference.getReference(): String? = when(this) {
    is SchemaReference.Resolved -> schema.ref?.substringAfterLast('/')
    is SchemaReference.Link.Array -> element.getReference()
    is SchemaReference.Link.Optional -> delegate.getReference()
    is SchemaReference.Link.Reference -> name
    is SchemaReference.Link.Simple -> null
}