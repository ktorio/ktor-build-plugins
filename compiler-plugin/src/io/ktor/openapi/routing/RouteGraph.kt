package io.ktor.openapi.routing

import org.jetbrains.kotlin.fir.FirSession

/**
 * Represents a graph of routes where a node can have multiple parents.
 */
class RouteGraph(val session: FirSession) {
    private val parentToChildren = mutableMapOf<RouteNode, MutableSet<RouteNode>>()
    private val childToParents = mutableMapOf<RouteNode, MutableSet<RouteNode>>()
    internal val routes = mutableListOf<RouteNode.Route>()
    internal val functions = mutableListOf<RouteNode.Function>()

    fun isEmpty() = routes.isEmpty()

    fun add(node: RouteNode) {
        for (i in routes.indices.reversed()) {
            val parent = routes[i]
            if (node in parent) {
                addEdge(parent, node)
                break
            } else if (parent.filePath != node.filePath || parent.isTopLevel()) {
                break
            }
        }
        when (node) {
            is RouteNode.Function -> functions.add(node)
            is RouteNode.Route -> routes.add(node)
            is RouteNode.CallFeature -> {}
        }
    }

    /**
     * Now that the function nodes are all accounted for, we can look for their children to
     * populate the edges.
     */
    fun build(): RouteGraph {
        for (function in functions) {
            for (other in functions) {
                if (other in function && other.isTopLevel())
                    addEdge(other, function)
            }
            for (route in routes) {
                if (route in function && route.isTopLevel())
                    addEdge(function, route)
            }
        }
        return this
    }

    fun RouteNode.isTopLevel(): Boolean =
        childToParents[this].orEmpty().none { it is RouteNode.Route }

    /**
     * Adds a parent-child relationship between nodes.
     */
    private fun addEdge(parent: RouteNode, child: RouteNode) {
        parentToChildren.getOrPut(parent) { mutableSetOf() }.add(child)
        childToParents.getOrPut(child) { mutableSetOf() }.add(parent)
    }

    /**
     * Checks if a node has any parents.
     */
    private fun hasParents(node: RouteNode): Boolean =
        childToParents.containsKey(node) && childToParents[node]?.isNotEmpty() == true

    /**
     * Gets all parents of a node.
     */
    private fun getParents(node: RouteNode): Set<RouteNode> =
        childToParents[node] ?: emptySet()

    /**
     * Gets all children of a node.
     */
    fun getChildren(node: RouteNode): Set<RouteNode> =
        parentToChildren[node] ?: emptySet()

    /**
     * Finds all paths from a node to root nodes (nodes with no parents).
     * Returns a list of paths, where each path is a list of nodes from the node to a root.
     */
    fun findAllPathsToRoots(node: RouteNode): List<List<RouteNode>> {
        if (!hasParents(node)) return listOf(listOf(node))

        val result = mutableListOf<List<RouteNode>>()
        val parents = getParents(node)

        for (parent in parents) {
            val parentPaths = findAllPathsToRoots(parent)
            for (path in parentPaths) {
                result.add(path + node)
            }
        }

        return result
    }
}