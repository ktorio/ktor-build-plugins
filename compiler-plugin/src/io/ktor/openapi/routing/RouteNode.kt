package io.ktor.openapi.routing

import io.ktor.compiler.utils.getArgument
import io.ktor.compiler.utils.getFunctionName
import io.ktor.compiler.utils.range
import org.jetbrains.kotlin.fir.expressions.FirFunctionCall
import org.jetbrains.kotlin.fir.expressions.FirLiteralExpression

/**
 * Code references to the Ktor routing API.  These recorded, then merged to form the model of the OpenAPI specification.
 */
sealed interface RouteNode: SourceCoordinates {
    val fir: FirFunctionCall
    val fields: RouteStack.() -> RouteFieldList

    val id: String get() = filePath.orEmpty() + startOffset.toString()
    val functionName: String get() = fir.getFunctionName()
    override val startOffset: Int get() = fir.source?.range?.first ?: -1
    override val endOffset: Int get() = fir.source?.range?.last ?: -1

    /**
     * Route call with method and path.
     *
     * These map to paths in the OpenAPI specification.
     */
    class Route(
        override val filePath: String?,
        override val fir: FirFunctionCall,
        override val fields: RouteStack.() -> RouteFieldList,
    ): RouteNode {
        val method: String? get() =
            functionName.takeIf { it in RoutingFunctionConstants.HTTP_METHODS }

        override fun toString(): String = buildString {
            append(fir.getFunctionName())
            (fir.getArgument("path") as? FirLiteralExpression)?.let {
                append(" ")
                append(it.value.toString())
            }
        }

        override fun equals(other: Any?): Boolean =
            other is Route && id == other.id

        override fun hashCode(): Int = id.hashCode()
    }

    /**
     * A general call to some Ktor feature that generates automatic OpenAPI properties.
     */
    class CallFeature(
        override val filePath: String?,
        override val fir: FirFunctionCall,
        override val fields: RouteStack.() -> RouteFieldList,
    ) : RouteNode {
        override fun equals(other: Any?): Boolean =
            other is CallFeature && id == other.id

        override fun hashCode(): Int = id.hashCode()
    }

    /**
     * Custom extension function of `Route`.
     *
     * Since it's a reference to a function, it may have non-local children.
     */
    class Function(
        override val filePath: String?,
        override val fir: FirFunctionCall,
        val declaration: SourceCoordinates,
        override val fields: RouteStack.() -> RouteFieldList,
    ): RouteNode {
        override fun contains(other: SourceCoordinates): Boolean =
            other in declaration && other != this

        override fun toString(): String =
            fir.getFunctionName() + '@' + fir.source?.range

        override fun equals(other: Any?): Boolean =
            other is Function && id == other.id

        override fun hashCode(): Int = id.hashCode()
    }
}