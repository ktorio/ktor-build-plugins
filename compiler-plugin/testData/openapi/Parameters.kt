// RUN_PIPELINE_TILL: BACKEND

package openapi

import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import kotlinx.serialization.*
import kotlinx.serialization.json.*

fun Application.parameters() {
    routing {
        /**
         * Basic parameter demonstration
         *
         * @path a [String] First path parameter
         * @path b [String] Second path parameter
         * @path c [String] Third path parameter
         * @query d [String] First query parameter
         * @query e [String] Second query parameter
         * @header f [String] First header parameter
         * @header g [String]+ Multiple header parameters with same name
         * @header h [String] Another header parameter
         * @response 200 All parameters as string
         */
        get("/parameters/{a}/{b}/{c}") {
            call.respondText(
                listOf(
                    call.parameters["a"],
                    call.pathParameters["b"],
                    call.request.pathVariables["c"],
                    call.queryParameters["d"],
                    call.request.queryParameters["e"],
                    call.request.headers["f"],
                    call.request.headers.getAll("g"),
                    call.request.header("h")
                ).joinToString()
            )
        }

        route("/typed-parameters") {
            /**
             * Integer parameters
             *
             * @path id [Int] Path ID
             *   minimum: 1
             *   maximum: 1000
             * @query limit [Int] Max number of items
             *   minimum: 1
             *   maximum: 100
             *   default: 20
             * @query page [Int] Page number
             *   minimum: 1
             *   default: 1
             * @response 200 [TypedResponse] Response with parsed values
             * @response 400 Invalid parameters
             */
            get("/int/{id}") {
                val id = call.parameters["id"]?.toIntOrNull()
                    ?: return@get call.respond(HttpStatusCode.BadRequest, "Invalid ID")

                if (id < 1 || id > 1000) {
                    return@get call.respond(HttpStatusCode.BadRequest, "ID must be between 1 and 1000")
                }

                val limit = call.request.queryParameters["limit"]?.toIntOrNull() ?: 20
                if (limit < 1 || limit > 100) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Limit must be between 1 and 100")
                }

                val page = call.request.queryParameters["page"]?.toIntOrNull() ?: 1
                if (page < 1) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Page must be at least 1")
                }

                call.respond(
                    TypedResponse(
                        id = id,
                        limit = limit,
                        page = page,
                        type = "integer"
                    )
                )
            }

            /**
             * Float parameters
             *
             * @path value [Double] Floating point value
             *   minimum: 0.0
             *   maximum: 100.0
             * @query factor [Float] Multiplication factor
             *   minimum: 0.1
             *   maximum: 10.0
             *   default: 1.0
             * @response 200 [TypedResponse] Response with parsed values
             * @response 400 Invalid parameters
             */
            get("/float/{value}") {
                val value = call.parameters["value"]?.toDoubleOrNull()
                    ?: return@get call.respond(HttpStatusCode.BadRequest, "Invalid value")

                if (value < 0.0 || value > 100.0) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Value must be between 0.0 and 100.0")
                }

                val factor = call.request.queryParameters["factor"]?.toFloatOrNull() ?: 1.0f
                if (factor < 0.1f || factor > 10.0f) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Factor must be between 0.1 and 10.0")
                }

                call.respond(
                    TypedResponse(
                        doubleValue = value,
                        floatValue = factor,
                        type = "float"
                    )
                )
            }

            /**
             * Boolean parameters
             *
             * @query enabled [Boolean] Feature flag
             *   default: false
             * @query verbose [Boolean] Verbose output
             *   default: false
             * @header debug [Boolean] Debug mode
             *   default: false
             * @response 200 [TypedResponse] Response with parsed values
             */
            get("/boolean") {
                val enabled = call.request.queryParameters["enabled"]?.toBoolean() ?: false
                val verbose = call.request.queryParameters["verbose"]?.toBoolean() ?: false
                val debug = call.request.header("debug")?.toBoolean() ?: false

                call.respond(
                    TypedResponse(
                        enabled = enabled,
                        verbose = verbose,
                        debug = debug,
                        type = "boolean"
                    )
                )
            }

            /**
             * String parameters with pattern constraints
             *
             * @path code [String] Product code
             *   pattern: ^[A-Z]{2}\d{4}$
             * @query email [String] User email
             *   pattern: ^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$
             * @query phone [String] Phone number
             *   pattern: ^\+\d{1,3}-\d{3,14}$
             * @response 200 [TypedResponse] Response with parsed values
             * @response 400 Invalid parameters
             */
            get("/pattern/{code}") {
                val code = call.parameters["code"]
                if (code == null || !Regex("^[A-Z]{2}\\d{4}$").matches(code)) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Invalid product code format")
                }

                val email = call.request.queryParameters["email"]
                if (email != null && !Regex("^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$").matches(email)) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Invalid email format")
                }

                val phone = call.request.queryParameters["phone"]
                if (phone != null && !Regex("^\\+\\d{1,3}-\\d{3,14}$").matches(phone)) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Invalid phone number format")
                }

                call.respond(
                    TypedResponse(
                        code = code,
                        email = email,
                        phone = phone,
                        type = "pattern"
                    )
                )
            }

            /**
             * Enum parameters
             *
             * @path status [String] Order status
             *   enum: [PENDING, PROCESSING, SHIPPED, DELIVERED, CANCELLED]
             * @query sort [String] Sort direction
             *   enum: [asc, desc]
             *   default: asc
             * @query filter [String] Result filter
             *   enum: [all, active, inactive]
             *   default: all
             * @response 200 [TypedResponse] Response with parsed values
             * @response 400 Invalid parameters
             */
            get("/enum/{status}") {
                val validStatuses = setOf("PENDING", "PROCESSING", "SHIPPED", "DELIVERED", "CANCELLED")
                val status = call.parameters["status"]
                if (status == null || status !in validStatuses) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Invalid status value")
                }

                val validSortDirections = setOf("asc", "desc")
                val sort = call.request.queryParameters["sort"] ?: "asc"
                if (sort !in validSortDirections) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Invalid sort direction")
                }

                val validFilters = setOf("all", "active", "inactive")
                val filter = call.request.queryParameters["filter"] ?: "all"
                if (filter !in validFilters) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Invalid filter value")
                }

                call.respond(
                    TypedResponse(
                        status = status,
                        sort = sort,
                        filter = filter,
                        type = "enum"
                    )
                )
            }

            /**
             * Array parameters
             *
             * @query ids [Int]+ List of IDs
             * @query tags [String]+ List of tags
             * @query flags [Boolean]+ List of flags
             * @response 200 [TypedResponse] Response with parsed arrays
             */
            get("/array") {
                val ids = call.request.queryParameters.getAll("ids")?.mapNotNull { it.toIntOrNull() } ?: emptyList()
                val tags = call.request.queryParameters.getAll("tags") ?: emptyList()
                val flags = call.request.queryParameters.getAll("flags")?.map { it.toBoolean() } ?: emptyList()

                call.respond(
                    TypedResponse(
                        ids = ids,
                        tags = tags,
                        flags = flags,
                        type = "array"
                    )
                )
            }

            /**
             * Date and time parameters
             *
             * @query date [String] Date in ISO format
             *   pattern: ^\d{4}-\d{2}-\d{2}$
             * @query time [String] Time in ISO format
             *   pattern: ^\d{2}:\d{2}(:\d{2})?$
             * @query datetime [String] DateTime in ISO format
             *   pattern: ^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{3})?(Z|[+-]\d{2}:\d{2})?$
             * @response 200 [TypedResponse] Response with parsed date/time values
             * @response 400 Invalid parameters
             */
            get("/datetime") {
                val date = call.request.queryParameters["date"]
                if (date != null && !Regex("^\\d{4}-\\d{2}-\\d{2}$").matches(date)) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Invalid date format")
                }

                val time = call.request.queryParameters["time"]
                if (time != null && !Regex("^\\d{2}:\\d{2}(:\\d{2})?$").matches(time)) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Invalid time format")
                }

                val datetime = call.request.queryParameters["datetime"]
                if (datetime != null && !Regex("^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d{3})?(Z|[+-]\\d{2}:\\d{2})?$").matches(
                        datetime
                    )
                ) {
                    return@get call.respond(HttpStatusCode.BadRequest, "Invalid datetime format")
                }

                call.respond(
                    TypedResponse(
                        date = date,
                        time = time,
                        datetime = datetime,
                        type = "datetime"
                    )
                )
            }
        }

        /**
         * Multiple parameters with the same name
         *
         * @query filter [String]+ Multiple filter parameters
         * @header x-custom [String]+ Multiple custom headers
         * @response 200 [MultiValueResponse] Response with all parameter values
         */
        get("/multi-value") {
            val filters = call.request.queryParameters.getAll("filter") ?: emptyList()
            val customHeaders = call.request.headers.getAll("x-custom") ?: emptyList()

            call.respond(
                MultiValueResponse(
                    filters = filters,
                    customHeaders = customHeaders
                )
            )
        }

        /**
         * Required vs Optional parameters
         *
         * @path id [Int] Required path parameter
         * @query required [String] Required query parameter
         * @query optional [String] Optional query parameter
         * @header x-required [String] Required header
         * @header x-optional [String] Optional header
         * @response 200 [RequiredOptionalResponse] Response with all parameters
         * @response 400 Missing required parameter
         */
        get("/required-optional/{id}") {
            val id = call.parameters["id"]?.toIntOrNull()
                ?: return@get call.respond(HttpStatusCode.BadRequest, "Missing required path parameter: id")

            val required = call.request.queryParameters["required"]
                ?: return@get call.respond(HttpStatusCode.BadRequest, "Missing required query parameter: required")

            val optional = call.request.queryParameters["optional"]

            val requiredHeader = call.request.header("x-required")
                ?: return@get call.respond(HttpStatusCode.BadRequest, "Missing required header: x-required")

            val optionalHeader = call.request.header("x-optional")

            call.respond(
                RequiredOptionalResponse(
                    id = id,
                    required = required,
                    optional = optional,
                    requiredHeader = requiredHeader,
                    optionalHeader = optionalHeader
                )
            )
        }

        /**
         * Parameters with default values
         *
         * @query page [Int] Page number
         *   default: 1
         * @query size [Int] Page size
         *   default: 20
         * @query sort [String] Sort field
         *   default: id
         * @query order [String] Sort order
         *   enum: [asc, desc]
         *   default: asc
         * @query filter [String] Filter
         *   default: all
         * @response 200 [DefaultValuesResponse] Response with default or provided values
         */
        get("/defaults") {
            val page = call.request.queryParameters["page"]?.toIntOrNull() ?: 1
            val size = call.request.queryParameters["size"]?.toIntOrNull() ?: 20
            val sort = call.request.queryParameters["sort"] ?: "id"
            val order = call.request.queryParameters["order"] ?: "asc"
            val filter = call.request.queryParameters["filter"] ?: "all"

            call.respond(
                DefaultValuesResponse(
                    page = page,
                    size = size,
                    sort = sort,
                    order = order,
                    filter = filter
                )
            )
        }

        /**
         * Combined parameter sources
         *
         * @path id [String] Resource ID
         * @query fields [String]+ Fields to include
         * @query expand [String]+ Related resources to expand
         * @header if-modified-since [String] Only return if modified since this date
         * @header x-api-key [String] API key for authentication
         * @response 200 [CombinedParametersResponse] Response with all parameters
         * @response 400 Invalid parameters
         * @response 401 Missing or invalid API key
         */
        get("/combined/{id}") {
            val id = call.parameters["id"]
                ?: return@get call.respond(HttpStatusCode.BadRequest, "Missing path parameter: id")

            val apiKey = call.request.header("x-api-key")
                ?: return@get call.respond(HttpStatusCode.Unauthorized, "Missing API key")

            if (apiKey != "valid-key") {
                return@get call.respond(HttpStatusCode.Unauthorized, "Invalid API key")
            }

            val fields = call.request.queryParameters.getAll("fields") ?: emptyList()
            val expand = call.request.queryParameters.getAll("expand") ?: emptyList()
            val ifModifiedSince = call.request.header("if-modified-since")

            call.respond(
                CombinedParametersResponse(
                    id = id,
                    fields = fields,
                    expand = expand,
                    ifModifiedSince = ifModifiedSince,
                    apiKey = apiKey
                )
            )
        }

        /**
         * Parameter encoding styles
         *
         * @query simple [String] Simple style parameter
         * @query array [String]+ Array of values
         * @query object [String] Object parameter (serialized as JSON)
         * @response 200 [EncodingStylesResponse] Response with parsed parameters
         */
        get("/encoding-styles") {
            val simple = call.request.queryParameters["simple"]
            val array = call.request.queryParameters.getAll("array") ?: emptyList()

            val objectParam = call.request.queryParameters["object"]
            val parsedObject = try {
                objectParam?.let { Json.decodeFromString<Map<String, String>>(it) }
            } catch (e: Exception) {
                null
            }

            call.respond(
                EncodingStylesResponse(
                    simple = simple,
                    array = array,
                    objectParam = parsedObject
                )
            )
        }

        /**
         * Header parameters with special handling
         *
         * @header user-agent [String] Client user agent
         * @header accept [String] Accepted content types
         * @header accept-language [String] Accepted languages
         * @header content-type [String] Content type of the request
         * @header authorization [String] Authorization header
         * @response 200 [HeaderParametersResponse] Response with parsed headers
         */
        get("/headers") {
            val userAgent = call.request.header("user-agent")
            val accept = call.request.header("accept")
            val acceptLanguage = call.request.header("accept-language")
            val contentType = call.request.header("content-type")
            val authorization = call.request.header("authorization")

            call.respond(
                HeaderParametersResponse(
                    userAgent = userAgent,
                    accept = accept,
                    acceptLanguage = acceptLanguage,
                    contentType = contentType,
                    authorization = authorization
                )
            )
        }

        /**
         * Path parameter with regex pattern
         *
         * @path uuid [String] UUID parameter
         *   pattern: ^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$
         * @response 200 UUID is valid
         * @response 400 Invalid UUID format
         */
        get("/uuid/{uuid}") {
            val uuid = call.parameters["uuid"]
            val uuidPattern = Regex("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")

            if (uuid == null || !uuidPattern.matches(uuid.lowercase())) {
                return@get call.respond(HttpStatusCode.BadRequest, "Invalid UUID format")
            }

            call.respondText("UUID is valid: $uuid")
        }

        /**
         * Cookie parameters
         *
         * @cookie session-id [String] Session identifier
         * @cookie preferences [String] User preferences
         * @response 200 [CookieParametersResponse] Response with parsed cookies
         */
        get("/cookies") {
            val sessionId = call.request.cookies["session-id"]
            val preferences = call.request.cookies["preferences"]

            call.respond(
                CookieParametersResponse(
                    sessionId = sessionId,
                    preferences = preferences
                )
            )
        }
    }
}

@Serializable
data class TypedResponse(
    val id: Int = 0,
    val limit: Int = 0,
    val page: Int = 0,
    val doubleValue: Double = 0.0,
    val floatValue: Float = 0.0f,
    val enabled: Boolean = false,
    val verbose: Boolean = false,
    val debug: Boolean = false,
    val code: String? = null,
    val email: String? = null,
    val phone: String? = null,
    val status: String? = null,
    val sort: String? = null,
    val filter: String? = null,
    val ids: List<Int> = emptyList(),
    val tags: List<String> = emptyList(),
    val flags: List<Boolean> = emptyList(),
    val date: String? = null,
    val time: String? = null,
    val datetime: String? = null,
    val type: String
)

@Serializable
data class MultiValueResponse(
    val filters: List<String>,
    val customHeaders: List<String>
)

@Serializable
data class RequiredOptionalResponse(
    val id: Int,
    val required: String,
    val optional: String?,
    val requiredHeader: String,
    val optionalHeader: String?
)

@Serializable
data class DefaultValuesResponse(
    val page: Int,
    val size: Int,
    val sort: String,
    val order: String,
    val filter: String
)

@Serializable
data class CombinedParametersResponse(
    val id: String,
    val fields: List<String>,
    val expand: List<String>,
    val ifModifiedSince: String?,
    val apiKey: String
)

@Serializable
data class EncodingStylesResponse(
    val simple: String?,
    val array: List<String>,
    val objectParam: Map<String, String>?
)

@Serializable
data class HeaderParametersResponse(
    val userAgent: String?,
    val accept: String?,
    val acceptLanguage: String?,
    val contentType: String?,
    val authorization: String?
)

@Serializable
data class CookieParametersResponse(
    val sessionId: String?,
    val preferences: String?
)