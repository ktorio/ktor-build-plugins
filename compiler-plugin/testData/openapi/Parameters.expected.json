{
    "openapi": "3.1.1",
    "info": {
        "title": "Open API Document",
        "version": "1.0.0"
    },
    "paths": {
        "/parameters/{a}/{b}/{c}": {
            "get": {
                "responses": {
                    "200": {
                        "description": "",
                        "content": {
                            "text/plain": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                },
                "parameters": [
                    {
                        "name": "a",
                        "in": "path",
                        "required": true,
                        "schema": {
                            "type": "string"
                        }
                    },
                    {
                        "name": "b",
                        "in": "path",
                        "required": true,
                        "schema": {
                            "type": "string"
                        }
                    },
                    {
                        "name": "c",
                        "in": "path",
                        "required": true,
                        "schema": {
                            "type": "string"
                        }
                    },
                    {
                        "name": "d",
                        "in": "query",
                        "required": false,
                        "schema": {
                            "type": "string"
                        }
                    },
                    {
                        "name": "e",
                        "in": "query",
                        "required": false,
                        "schema": {
                            "type": "string"
                        }
                    },
                    {
                        "name": "f",
                        "in": "header",
                        "required": false,
                        "schema": {
                            "type": "string"
                        }
                    },
                    {
                        "name": "g",
                        "in": "header",
                        "required": false,
                        "schema": {
                            "type": "string"
                        }
                    },
                    {
                        "name": "h",
                        "in": "header",
                        "required": false,
                        "schema": {
                            "type": "string"
                        }
                    }
                ]
            }
        }
    },
    "components": {
        "schemas": {}
    }
}